### PR TITLE
Exclude COMMAND.COM's help arguments when part of sub-command (#2263)

### LINKED ISSUE #2263: `/?` command line parameter handling weirdness
1. Execute any built-in DOS command with the `/?` parameter (e.g. `DIR /?`), or any 3rd party executable that supports the `/?` param to print out some built-in help (e.g. `DOSMID /?`)
2. Observe that the help for the invoked command is printed out.
3. Start DOS Navigator, then execute the same `SOMEEXE /?` command.
4. Note that regardless of the invoked command, the help for the built-in DOS command shell is printed out, without exception (same as invoking `COMMAND /?`)

### PR DESCRIPTION
**Pull request by kcgen**
Fixes #2263

### CODE REVIEW

File: src/misc/setup.cpp [hunk 1]
```diff
@@ -1234,7 +1234,33 @@ bool CommandLine::FindExist(const char *name, bool remove)
 	return true;
 }
 
-bool CommandLine::FindInt(const char *name, int &value, bool remove)
+// Checks if any of the command line arguments are found in the anterior
+// arguments and come before any of the posterior arguments. If none of the
+// command line arguments are found in the anterior then false is returned.
+bool CommandLine::ExistsPriorTo(const std::list<std::string>& anterior,
+                                const std::list<std::string>& posterior) const
```
Comment (johnnovak) [inline src/misc/setup.cpp:1241]: minor (and only a preference 😄): maybe call these `pre_args` and `post_args`?

File: src/misc/setup.cpp [hunk 2]
```diff
@@ -1234,7 +1234,33 @@ bool CommandLine::FindExist(const char *name, bool remove)
...
+	                         const std::string& needle) {
+		return std::any_of(haystack.begin(),
+		                   haystack.end(),
+		                   [&needle](const std::string& hay) {
+			                   return iequals(hay, needle);
+		                   });
+	};
+
+	for (const auto& cmd : cmds) {
+		if (any_of_iequals(anterior, cmd)) {
```
Comment (johnnovak) [inline src/misc/setup.cpp:1253]: `cmd` is not really a "command" but rather an `arg`(ument), isn't it?

File: src/shell/shell.cpp [hunk 3]
```diff
@@ -489,14 +489,17 @@ void DOS_Shell::RunInternal()
 
 void DOS_Shell::Run()
 {
-	char input_line[CMD_MAXLINE] = {0};
-	std::string line;
-	if (cmd->FindExist("/?", false) || cmd->FindExist("-?", false)) {
+	// COMMAND.COM's /C and /INIT spawn sub-commands. When parsing help, we need
+	// to be sure the /? and -? are intended for us and not part of the
+	// sub-command.
+	if (cmd->ExistsPriorTo({"/?", "-?"}, {"/C", "/INIT"})) {
```
Comment (johnnovak) [inline src/shell/shell.cpp:495]: Nifty fix! And now the behaviour makes complete sense! 😅

**Review by johnnovak — APPROVED** [review]
Looks good @kcgen, that was a quick fix! 🎉

File: src/misc/setup.cpp (hunk 1 above)
Comment (kcgen) [reply src/misc/setup.cpp]: Yes; much better 👍!

File: src/misc/setup.cpp (hunk 2 above)
Comment (kcgen) [reply src/misc/setup.cpp]: I agree; but I was following the original design. It starts with the class "CommandLine" which stores all the CLI insides a "cmds" std::list member variable, which this code is iterating over.

I like to use `for (const auto& singular : plural)`, in this case `cmd` and `cmds`.

But yeah.. cmds would be better named as args. 

So I've named the singular `cli_arg`, just to make it more clear that this is the actual arg from the command like, where as the pre and post args are just canned sets for comparison.

So

File: src/shell/shell.cpp (hunk 3 above)
Comment (kcgen) [reply src/shell/shell.cpp:495]: Thanks for the quick check and review, @johnnovak !

**Comment by kcgen** [comment]
The PR now makes the `iequals(a, b)` case-insensitive comparison a constexpr template that can handle mixed types of `const char*`, `const std::string&`, and `std::string_view`. 
 
The benefit is that any "text" literals  passed in become string_views, and won't have to be copied into std::string objects. 

This avoids allocating and deallocating the string temporary, and avoids copying value by value into the string. 

Not that these operations are perfomance critical, but might as well lighten the load.

**Pull request merged**